pkg/cdi/spec.go

```diff
@@ -109,6 +110,56 @@ func NewSpec(raw *cdi.Spec, path string, priority int) (*Spec, error) {
 	return spec, nil
 }
 
+// Write the CDI Spec to the file associated with it during instantiation
+// by NewSpec() or ReadSpec().
+func (s *Spec) Write(overwrite bool) error {
+	var (
+		data []byte
+		dir  string
+		tmp  *os.File
+		err  error
+	)
+
+	err = validateSpec(s.Spec)
+	if err != nil {
+		return err
+	}
+
+	if filepath.Ext(s.path) == ".yaml" {
+		data, err = yaml.Marshal(s.Spec)
+	} else {
+		data, err = json.Marshal(s.Spec)
+	}
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal Spec file")
+	}
+
+	dir = filepath.Dir(s.path)
+	err = os.MkdirAll(dir, 0o755)
+	if err != nil {
+		return errors.Wrap(err, "failed to create Spec dir")
+	}
+
+	tmp, err = os.CreateTemp(dir, "spec.*.tmp")
+	if err != nil {
+		return errors.Wrap(err, "failed to create Spec file")
+	}
+	_, err = tmp.Write(data)
+	tmp.Close()
+	if err != nil {
+		return errors.Wrap(err, "failed to write Spec file")
+	}
+
+	err = renameAt(dir, filepath.Base(tmp.Name()), filepath.Base(s.path), overwrite)
```

Comment:
Looking at this again: I have two comments. 
1. The arguments of the function `renameAt` here doesn't match those for the `RenameAt` or `RenameAt2` calls used to do the actual rename
2. Using the dir as an argument seems to be more cumbersome than just using filenames. (we can still extract the directories as required in the `rename` implementation).

Reply:
The current function signature is not an accident. We do want to enforce the two files to already be in the same directory to make sure they are on the same mounted device/filesystem. Otherwise `renameat2()` would fail. If the differing signature wrt. renameat{,2} is too disturbing, I'd suggest we call it `ReplaceAt()` instead, with the same signature as now.

Reply:
OK. The check for the same device / filesystem is important in this case. I think `replaceAt` implies an overwrite, so what about `renameIn` instead? From my point of view this makes it clear that the operation is happening in a specified directory with the two files passed as arguments.

I'm also ok to leave it as is if you're strongly against `renameIn` as an option.

Reply:
You are right. `renameIn()` is better, because replacing is behind a boolean flag. I'll update accordingly.